pkgs/development/tools/godot/3/default.nix

```diff
@@ -1,109 +1,165 @@
-, openssl
-, alsa-lib
-, alsa-plugins
 , makeWrapper
-, libGLU
-, zlib
+, openssl
+, pkg-config
+, scons
+, udev
 , yasm
-, withUdev ? true
+, zlib
 }:
 
-let
-  options = {
-    touch = libXi != null;
-    pulseaudio = false;
-    udev = withUdev;
-  };
-in
-stdenv.mkDerivation rec {
-  pname = "godot";
+stdenv.mkDerivation (self: {
+  pname = "godot3";
   version = "3.5.2";
+  buildDescription = "X11 tools";
 
   src = fetchFromGitHub {
     owner = "godotengine";
     repo = "godot";
-    rev = "${version}-stable";
+    rev = "${self.version}-stable";
     sha256 = "sha256-C+1J5N0ETL1qKust+2xP9uB4x9NwrMqIm8aFAivVYQw=";
   };
 
-  nativeBuildInputs = [ pkg-config makeWrapper ];
-  buildInputs = [
+  nativeBuildInputs = [
+    installShellFiles
+    pkg-config
+    makeWrapper
     scons
-    udev
+  ];
+
+  buildInputs = [
+    alsa-lib
+    freetype
+    libGLU
+    libpulseaudio
     libX11
     libXcursor
+    libXext
+    libXfixes
+    libXi
     libXinerama
     libXrandr
     libXrender
-    libXi
-    libXext
-    libXfixes
-    freetype
     openssl
-    alsa-lib
-    libpulseaudio
-    libGLU
-    zlib
+    udev
     yasm
+    zlib
   ];
 
-  patches = [ ./pkg_config_additions.patch ./dont_clobber_environment.patch ];
+  shouldAddLinkFlagsToPulse = true;
+
+  patches = map (rp: ./patches + rp) (
+    [
+      # The version of SConstruct in the godot source appends the OS's PATH to the Scons PATH,
+      # but because it is an append, the Scons PATH takes precedence.  The Scons PATH contains a
+      # bunch of standard Linux paths like /usr/bin, so if they happen to contain versions of any
+      # build-time dependencies of Godot, they will be used instead of the Nix version of them.
+      #
+      # This patch simply replaces the entire Scons environment (including the PATH) with that
+      # of the OS. This isn't as surgical as just fixing the PATH, but it seems to work, and
+      # seems to be the Nix community's current strategy when using Scons.
+      /SConstruct/dontClobberEnvironment.patch
+
+      # The version of detect.py in the godot source specifies both the compile flags and link flags
+      # for most of godot's dependencies, but for alsa, pulseaudio, and udev, it omits the link flags.
+      # It has comments by them explaining that the link flags are not needed because godot will dlopen
+      # these libraries. However, the dlopen call seems to fail for the Nix-based builds, and including
+      # the link flags seems to fix it. This patch does just that by passing `--libs` to pkg-config.
+      /detect.py/addLinkFlagsToAlsaAndUdev.patch
+    ] ++
+
+    # Adding link flags to pulseaudio is necessary to get the godot tool to work, but it seems to break
+    # the export templates. This doesn't seem to apply to udev or alsa for some reason. It is made
+    # optional so the export templates derivation can omit the link flags.
+    lib.optional self.shouldAddLinkFlagsToPulse /detect.py/addLinkFlagsToPulse.patch
```

Comment:
Would just doing a `patchelf --add-needed` in `installPhase` work for how they `dlopen` it instead of modifying the link flags for the package?

I'm not too familiar with the situation behind this, but especially since one of these patches is conditional it would be nice if we can resolve this without them

Reply:
I've got no idea, not really familiar; I'll try it out and see if that works.  Thanks for the suggestion!

Reply:
I don't understand this stuff very well, but here's what I'm seeing...

To keep the testing simple, I played with just the unconditional patch: addLinkFlagsToAlsaAndUdev.patch.  With this patch, the godot3 application runs without error.  Without the patch, the terminal displays:
```
libasound.so.2: cannot open shared object file: No such file or directory
libudev.so.1: cannot open shared object file: No such file or directory
```
In spite of these errors, the application seems to run fine (I'm not sure what impact the errors actually have).  This what I was seeing before.

Then I applied this to the built binary: `patchelf --add-needed libasound.so.2 --add-needed libudev.so.1 <binaryname>`.

The resulting program gives this error and entirely fails to start:
```
error while loading shared libraries: libasound.so.2: cannot open shared object file: No such file or directory
```

Did I do something wrong or is there more to it?

Reply:
Hmm, maybe you'd need to add `${lib.getLib alsa-lib}/lib/libasound.so.2` and `${systemdMinimal}/lib/libudev.so.1` instead of just the lib name alone

But if it doesn't work even with that either, then it's probably not worth persuing further in this PR and just dealing with it in a follow-up if someone figures something out

Reply:
Okay, I've discovered some things.

1) Using `patchelf` with explicit paths worked for the godot tool itself.  If it's applied to the export template then programs built from it, such as pixelorama, break.  (In fact, it seems we can't `patchelf` the export template *at all* without breaking things built from it.)
2) Using `patchelf` with just library file names works just as well if you add `autoPatchelfHook` to the `nativeBuildInputs` and ensure any required libraries are in the  `buildInputs`.  (This also breaks the export template of course.)
3) This can be further simplified by removing the call to `patchelf` entirely, keeping `autoPatchelfHook` in the `nativeBuildInputs`, and then adding this attribute to the godot derivation: `runtimeDependencies = map lib.getLib [ alsa-lib libpulseaudio udev ]`.  This is noticed by `autoPatchelf`, and causes these libraries to be unconditionally added to rpath of the executable, according to the nixpkgs manual.  This works because these three libraries are used via dlopen.  (But again, this breaks the export template and derived games.)
4) If I completely disable any `patchelf`-ing, whether manual or via `autoPatchelfHook`, for the export template, then games derived from it stop being _completely_ broken and just go back to being unable to open those libraries.  I can then resolve that by adding `autoPatchelfHook` and `nativeBuildInputs` directly to the derivation of the game itself.  (So the patching occurs *after* the game has been built from the export template.)  With this in place it works.
5) Thus I can get a working solution by having the godot derivation conditionally patch the binary when building the tool, not the export template, and then update all the game derivations to patch their own binaries with their dependencies.

So we're left with a decision between two options:
1) Continue using a source patch.  This is more complicated than the patchelf approach from the perspective of the godot derivation, but doesn't require any patching from the derivations of the games built upon it.
2) Switch to patching the binary, and update all the game derivations to handle their own patching.

Simplifying the derivations of the individual game derivations is a strong incentive for option 1, but I have some concerns with it lead me to favoring option 2:
1) The source patch approach solves the problem by (I believe) causing the libraries to be statically linked into the executable.  Having them as runtime dependencies is better.
2) The source patch approach is inconsistent and mysterious:  The godot tool seems to need alsa, udev, and pulseaudio to be statically linked.  The export template only needs alsa and udev, and seems to break if I statically link pulseaudio.  It is uncertain why this happens, and for all I know, some other game could require a different subset to be statically linked to the export template.

So while it means a bit more work in the game derivations, I think it's a cleaner and more reliable approach.  Thoughts?

Reply:
Isn't there also option three leaving it as is with the current patch and not using patch-elf?

Reply:
That's what I mean by option 1, the source patch.  Basically need to decide between leaving it the way it is, and switching to patching the binary, and my inclination is to switch.

Reply:
oh sorry :see_no_evil:
I guess option two would be cleaner but harder for new packages so we should probably add a godot section to `doc/languages-frameworks/` explaining how to package a godot application.

Reply:
I'd mostly be inferring how to do it from the existing examples (there are three).  My only contribution to the game derivations would be the patchelfing.  And there are a lot of non-godot derivations that seem to do the same thing.  I'd rather leave the kind of documentation you describe for someone who's actually built a godot game derivation.

Reply:
Well if there aren't any objections, I'll go ahead and work on switching to the patchelfing approach tonight.

Reply:
> 1. The source patch approach solves the problem by (I believe) causing the libraries to be statically linked into the executable. Having them as runtime dependencies is better.

As far as I understand, it's still dynamic linking, but it forces them to be linked directly instead of `dlopen`'d at runtime

> 2. The source patch approach is inconsistent and mysterious: The godot tool seems to need alsa, udev, and pulseaudio to be statically linked. The export template only needs alsa and udev, and some games seem to break if I statically link pulseaudio. It is uncertain why this happens, and for all I know, some other game could require a different subset to be statically linked to the export template.

It does suck a bit that games will all have to manually handle patching themselves, so there is clearly a drawback to each approach, so I'll defer to your judgement on which is better

I personally would lead slightly more to the one that "automatically" handles more cases if it is adequately documented what and why it is happening. Perhaps you could just conditionally use patchelf for pulse specifically, though?

It just seems super weird and smells like an upstream bug that adding pulseaudio as a needed .so for games that do ostensibly need it causes them to crash (and I imagine they may be doing some shenanigans to either resolve it at runtime or are just using Pulse's ALSA frontend)

Reply:
> It just seems super weird and smells like an upstream bug that adding pulseaudio as a needed .so for games that do ostensibly need it causes them to crash (and I imagine they may be doing some shenanigans to either resolve it at runtime or are just using Pulse's ALSA frontend)

Yeah, I have no idea...  I don't see what's special/different about pulseaudio, other than just the result of empirically testing the results of patching/not-patching it.  This is part of why I'm concerned about the source patching approach:  I don't really understand why it needs to be conditional for pulseaudio and not the others, and for all I know, the others might need to be conditional as well for some games.  Different godot games use different subsets of these libraries (I think it just never dlopens pulseaudio if the specific game doesn't utilize it).

> It does suck a bit that games will all have to manually handle patching themselves, so there is clearly a drawback to each approach, so I'll defer to your judgement on which is better
> 
> I personally would lean slightly more to the one that "automatically" handles more cases if it is adequately documented what and why it is happening. Perhaps you could just conditionally use patchelf for pulse specifically, though?

I'm still leaning towards the patchelf approach because it feels like it gives us more consistent and predictable results.  While having the patching on a per-game basis is less than ideal, I think this could be resolved in the future by defining a `mkGodotDerivation` function (or something) that encapsulates the patchelfing, and define all the games' derivations with it.  Though, since different games use different subsets, we don't always have to patchelf all these libraries for every game.